File: Tree/472-getTargetCopy-dfs.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def getTargetCopy(self, original, cloned, target):
        """
        :type original: TreeNode
        :type cloned: TreeNode
        :type target: TreeNode
        :rtype: TreeNode
        """
        def dfs2(a,b,t):
            if not a:
                return
            if a is t:
                return b
            x=dfs2(a.left,b.left,t)
            if x:
                return x
            return dfs2(a.right,b.right,t)
        self.ans=None  #why self.ans is okay
        def  dfs(a,b,t):
            if not a:
                return 
            if a is t:
                self.ans=b
                return
            dfs(a.left,b.left,t)
            dfs(a.right,b.right,t)
        dfs(original,cloned,target)
        return self.ans
      

File: Tree/472-getTargetCopy-dfs.cpp
/**
 * @file 472-getTargetCopy-dfs.cpp
 * 
 * @author your name (you@domain.com)
 * @brief 
 * @version 0.1
 * @date 2024-04-03
 * 
 * @copyright Copyright (c) 2024
 * Given two binary trees original and cloned and given a reference to a node target in the original tree.

The cloned tree is a copy of the original tree.

Return a reference to the same node in the cloned tree.

Note that you are not allowed to change any of the two trees or the target node and the answer must be a reference to a node in the cloned tree.
 */
#include<iostream>

class TreeNode
{
    public:
        int m_val{};
        TreeNode* left{};
        TreeNode* right{};
        TreeNode(int v,TreeNode* l=nullptr,TreeNode* r=nullptr):m_val{v},left{l},right{r}{};
};



TreeNode* foo(TreeNode* a,TreeNode* b, TreeNode* target)
{
    if(!a)
        return nullptr;
    if(a->m_val==target->m_val)
        return b;
    foo(a->left,b->left,target);
    foo(a->right,b->right,target);
}




TreeNode* getTargetCopy(TreeNode* a,TreeNode* b, TreeNode* target)
{
    return foo(a,b,target);
}

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        // return bfs(cloned,target);
        if(original==nullptr||original==target)
            return cloned;
        
        auto left=getTargetCopy(original->left,cloned->left,target);
        if(left)
            return left;
        return getTargetCopy(original->right,cloned->right,target);
    }
private:
    // TreeNode* preOrder(TreeNode* root, TreeNode* target)
    // {
    //     if(!root)
    //         return nullptr;
    //     if(root->val==target->val)
    //         return root;
    //     TreeNode* left= preOrder(root->left,target);
    //     TreeNode* right= preOrder(root->right,target);
    //     return left?left:right;
    // }
    // TreeNode* bfs(TreeNode* root, TreeNode* target)
    // {
    //     queue<TreeNode*> q;
    //     q.push(root);
    //     while(!q.empty()){
    //         TreeNode* curNode=q.front();
    //         q.pop();
    //         if(curNode->val==target->val)
    //             return curNode;
    //         if(curNode->left)
    //             q.push(curNode->left);
    //         if(curNode->right)
    //             q.push(curNode->right);
    //     }
    //     return nullptr;
    // }
};